### ISSUE TITLE
Maximum call stack size exceeded Exception

### ISSUE THREAD

**Issue by amcdnl** [theKashey/react-focus-lock#35]
I'm getting the exception below when I have a focus trapped dialog and a user clicks a button to open another dialog that is not focus trapped.

```
Uncaught RangeError: Maximum call stack size exceeded.
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
```

and 

```
Uncaught Error: An error was thrown inside one of your components, but React doesn't know what it was. This is likely due to browser flakiness. React does its best to preserve the "Pause on exceptions" behavior of the DevTools, which requires some DEV-mode only tricks. It's possible that these don't work in your browser. Try triggering the error in production mode, or switching to a modern browser. If you suspect that this is actually an issue with React, please file an issue.
    at Object.invokeGuardedCallbackDev (react-dom.development.js:143)
    at Object.invokeGuardedCallback (react-dom.development.js:187)
    at Object.invokeGuardedCallbackAndCatchFirstError (react-dom.development.js:201)
    at executeDispatch (react-dom.development.js:461)
    at executeDispatchesInOrder (react-dom.development.js:480)
    at executeDispatchesAndRelease (react-dom.development.js:581)
    at executeDispatchesAndReleaseTopLevel (react-dom.development.js:592)
    at forEachAccumulated (react-dom.development.js:562)
    at runEventsInBatch (react-dom.development.js:723)
    at runExtractedEventsInBatch (react-dom.development.js:732)
    at handleTopLevel (react-dom.development.js:4477)
    at batchedUpdates$1 (react-dom.development.js:16660)
    at batchedUpdates (react-dom.development.js:2131)
    at dispatchEvent (react-dom.development.js:4556)
    at interactiveUpdates$1 (react-dom.development.js:16715)
    at interactiveUpdates (react-dom.development.js:2150)
    at dispatchInteractiveEvent (react-dom.development.js:4533)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
    at HTMLDocument.onTrap (Trap.js:74)
    at HTMLDocument.Modal._this.enforceFocus (Modal.js:191)
    at focusOn (setFocus.js:15)
    at webpackJsonp../node_modules/focus-lock/dist/setFocus.js.exports.default (setFocus.js:25)
    at activateTrap (Trap.js:63)
```

**Comment by theKashey** [comment]
This is not a bug, this is a feature.
Look like Modal has got it's own focus management system and it results two active locks on the same page. Both of them are doing the job well, returning the focus to the lock, endlessly fighting each other.

There are 2 ways how to handle this situation:
 - disable FocusLock on modal open, it has a special prop for it (not easy to do)
 - wrap modal with FocusLock - it should deactivate current active Lock, and properly handle focus inside modal (not always possible)
 - easiest way - set `group` prop on FocusLock, and wrap modal, or modal internals with any element with `data-focus-lock="groupName"` attribute set. FocusLock will threat modal as a part of itself. Not preventing focus from being there.

Probably I should expose one more component, to "hide" some locations from FocusLock, as long this "fighting" is a common thing :(

**Comment by amcdnl** [comment]
> This is not a bug, this is a feature.

Ya, after looking at it a bit more I kinda realized this. I do think there is an opportunity for it to fail gracefully though. Since its getting in a circular loop, perhaps check what the current active focus item is and if its the same as what you are about to focus, don't refocus. Thoughts?

**Comment by theKashey** [comment]
1. Detect and break loop
2. Expose special component, to hide some DOM elements from FocusLock.

**Comment by amcdnl** [comment]
This is a project I worked on for Angular that does something similar to this lib - might be worth a look at how they do it - https://github.com/angular/material2/tree/master/src/cdk/a11y/focus-trap

**Comment by theKashey** [comment]
The problem is simple - lock should be very aggressive and should be very selfish - different locks are like spiders in a can - will try to eat each other.

**Comment by theKashey** [comment]
Solved in v1.12.0
